### PR TITLE
PXC-4341: Executing prepared statement can abort node after FLUSH TABLES

### PR DESCRIPTION
**Pull request by venkatesh-prasad-v**
https://perconadev.atlassian.net/browse/PXC-4341

Problem
-------
Transient errors during the execution of the DDL in a prepared statement mode can trigger inconsistency voting resulting in node eviction from the cluster.

Analysis & Fix
--------------
It can sometimes happen that the DDL executed as a prepared statement may fail with transient errors like ER_NEED_REPREPARE if the table metadata was changed between PREPARE and EXECUTE (due to FLUSH TABLES). In such cases, the statement is re-prepared and re-executed from Prepared_statement::execute_loop().

However, before the statement is reprepared, PXC's TOI_end() hook captures this transient error (i.e, ER_NEED_REPREPARE) and sends it to galera thereby triggering the inconsistenct voting and kicking this node out of the cluster.

So, in such cases, we need to delay sending the final error to galera until all reprepare attempts are over. This is done by attaching a temporary Diagnostics_area to THD and resetting it later if all the required conditions are met.

NOTE
----

The below mentioned commands reprepare the statement if there is any
metadata change (marked with CF_REEXECUTION_FRAGILE)
    
        SQLCOM_CREATE_TABLE
        SQLCOM_LOAD
        SQLCOM_CREATE_VIEW
        SQLCOM_UPDATE
        SQLCOM_UPDATE_MULTI
        SQLCOM_INSERT
        SQLCOM_INSERT_SELECT
        SQLCOM_DELETE
        SQLCOM_DELETE_MULTI
        SQLCOM_REPLACE
        SQLCOM_REPLACE_SELECT
        SQLCOM_SELECT
        SQLCOM_DO
        SQLCOM_CALL
    
Among the above commands, only SQLCOM_CREATE_TABLE and
SQLCOM_CREATE_VIEW are DDLs and thus was affected by this bug.

However in case of CREATE VIEW, table version check is performed before
the query is replicated as TOI, so CREATE TABLE was the only query that
was affected by this bug. So, a special handling has been done only for
the CREATE_TABLE query in this patch.

Testing Done
----
Jenkins: https://pxc.cd.percona.com/view/8.0%20parallel%20MTR/job/pxc-8.0-param-parallel-mtr/180/console

### CODE REVIEW

**Review by kamil-holubicki — APPROVED** [review]
lgtm

File: sql/sql_cmd_ddl_table.cc [hunk 1]
```diff
@@ -516,11 +516,16 @@ bool Sql_cmd_create_table::execute(THD *thd) {
...
-        cleanup action on TOI failure. */
-        if (WSREP(thd) && wsrep_to_isolation_begin(thd, create_table->db,
-                                                   create_table->table_name,
-                                                   NULL, NULL, &alter_info)) {
+        bool prepared_stmt_execution = thd->get_reprepare_observer() != nullptr;
+        bool should_start_toi = !prepared_stmt_execution ||
+                                (prepared_stmt_execution &&
+                                 !wsrep_thd_is_in_to_isolation(thd, false));
+        /* Note we are explicitly opening the macro as we need to perform
+        cleanup action on TOI failure. */ // KH: here TOI begin
```
Comment (kamil-holubicki) [inline sql/sql_cmd_ddl_table.cc]: Please remove 'KH:' comment

File: sql/sql_parse.cc [hunk 2]
```diff
@@ -6093,10 +6093,60 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
+      } else {
+        // This means all retries are exhausted. We should call TOI_end now.
+        finish_toi = true;
+      }
+    }
+  }
   thd->wsrep_consistency_check = NO_CONSISTENCY_CHECK;
-  WSREP_TO_ISOLATION_END;
+  if (finish_toi) {
+    WSREP_TO_ISOLATION_END;  // KH: here toi end
```
Comment (kamil-holubicki) [inline sql/sql_parse.cc]: Please remove 'KH:' comment

**Review by kamil-holubicki — CHANGES_REQUESTED** [review]

File: sql/sql_cmd_ddl_table.cc (hunk 1 above)
Comment (venkatesh-prasad-v) [reply sql/sql_cmd_ddl_table.cc]: Done

File: sql/sql_parse.cc (hunk 2 above)
Comment (venkatesh-prasad-v) [reply sql/sql_parse.cc]: Done

**Review by kamil-holubicki — APPROVED** [review]

**Pull request merged**